src/base/emu-i386/kvm.c

```diff
@@ -158,6 +158,76 @@ static struct kvm_userspace_memory_region maps[MAXSLOT];
 
 static int init_kvm_vcpu(void);
 
+#if !defined(DISABLE_SYSTEM_WA) || !defined(KVM_CAP_IMMEDIATE_EXIT)
+
+/* compat functions for older complex method of immediate exit
+   using a signal that is blocked outside KVM_RUN but not blocked inside it */
+
+#include <pthread.h>
+
+#ifndef KVM_CAP_IMMEDIATE_EXIT
+#define KVM_CAP_IMMEDIATE_EXIT 136
+#define immediate_exit padding1[0]
+#endif
+
+#define KVM_IMMEDIATE_EXIT_SIG (SIGRTMIN + 1)
+static int kvm_cap_immediate_exit = -1;
+
+static void kvm_compat_set_immediate_exit(int set)
+{
+  if (kvm_cap_immediate_exit == -1) { // setup
+    struct sigaction sa;
+    sigset_t set;
+    struct kvm_signal_mask *mask;
+
+    kvm_cap_immediate_exit = ioctl(kvmfd, KVM_CHECK_EXTENSION, KVM_CAP_IMMEDIATE_EXIT) > 0;
+    if (kvm_cap_immediate_exit)
+      return;
+
+    sa.sa_flags = 0;
+    sigemptyset(&sa.sa_mask);
+    sa.sa_handler = SIG_DFL; // never called
+    sigaction(KVM_IMMEDIATE_EXIT_SIG, &sa, NULL);
+
+    /* block KVM_IMMEDIATE_EXIT_SIG in main thread */
+    sigemptyset(&set);
+    sigaddset(&set, KVM_IMMEDIATE_EXIT_SIG);
+    pthread_sigmask(SIG_BLOCK, &set, NULL);
+
+    /* don't block any signals inside the guest */
+    sigemptyset(&set);
+    mask = malloc(sizeof(*mask) + sizeof(set));
+    mask->len = 8;
+    memcpy(mask->sigset, &set, sizeof(set));
```

Comment:
I believe it would be better to make
`set` a pointer that points to mask->sigset,
or just directly use `mask->sigset` in all
operations, removing `set`.
Also perhaps `alloca()` is a good candidate
for allocating that discrepancy.

Reply:
I could eliminate the dynamical allocation using a struct in the end.
It's an issue between glibc's idea of sigset_t and the kernel's idea, and the struct defintion:
```
/* for KVM_SET_SIGNAL_MASK */
struct kvm_signal_mask {
        __u32 len;
        __u8  sigset[0];
};
```
looking at the kernel code KVM_SET_SIGNAL_MASK errors unless len == 8.
So copying the first 8 bytes of a glibc sigset_t to the sigset above does the trick, as long as at least 7 bytes of padding is added after.

Reply:
super-crappy interface for sure... at least it's not used much any more